wallets/provider-solflare-snap/src/signers/solanaSigner.ts

```diff
@@ -0,0 +1,37 @@
+import type SolflareMetaMask from '@solflare-wallet/metamask-sdk';
+import type { GenericSigner, SolanaTransaction } from 'rango-types/lib';
+
+import {
+  getSolanaConnection,
+  prepareTransaction,
+} from '@rango-dev/signer-solana';
+import { SignerError, SignerErrorCode } from 'rango-types';
+
+export class SolflareSnapSolanaSigner
+  implements GenericSigner<SolanaTransaction>
+{
+  private provider: SolflareMetaMask;
+
+  constructor(provider: SolflareMetaMask) {
+    this.provider = provider;
+  }
+
+  async signMessage(): Promise<string> {
+    throw SignerError.UnimplementedError('signMessage');
+  }
+
+  async signAndSendTx(tx: SolanaTransaction): Promise<{ hash: string }> {
+    try {
+      const connection = getSolanaConnection();
+      const latestBlock = await connection.getLatestBlockhash('confirmed');
+
+      const finalTx = prepareTransaction(tx, latestBlock.blockhash);
+
+      const hash = await this.provider.signAndSendTransaction(finalTx);
```

Comment:
Please refactor and export this logic in solana-signer and call it here:

```ts
async function simulate (tx) {
  // We first simulate whether the transaction would be successful
  if (tx.txType === 'VERSIONED') {
    const { value: simulatedTransactionResponse } =
      await connection.simulateTransaction(
        finalTx as VersionedTransaction,
        {
          replaceRecentBlockhash: true,
          commitment: 'processed',
        }
      );
    const { err, logs } = simulatedTransactionResponse;
    if (err) {
      /*
       * Simulation error, we can check the logs for more details
       * If you are getting an invalid account error, make sure that you have the input mint account to actually swap from.
       */
      console.error('Simulation Error:', { err, logs });
      const message =
        (logs?.length || 0) > 0
          ? logs?.[logs?.length - 1]
          : JSON.stringify(err);
  
      throw new SignerError(
        SignerErrorCode.SEND_TX_ERROR,
        undefined,
        `Simulation failed: ${message}`
      );
    }
  }
}
```

and call it here:

```suggestion
      await simulateTransaction(finalTx);
      const hash = await this.provider.signAndSendTransaction(finalTx);
```

also you need to update your catch clause like this:

```ts
    } catch (error) {
      if (error instanceof SignerError) throw error;
      throw new SignerError(SignerErrorCode.SEND_TX_ERROR, undefined, error);
    }
```